drivers/dma/dma_rpi_pico.c

```diff
@@ -0,0 +1,409 @@
+	    dma_cfg->source_data_size != 4) {
+		LOG_DBG("source_data_size must be 1, 2, or 4 (%" PRIu32 ")",
+			dma_cfg->source_data_size);
+		return -EINVAL;
+	}
+
+	if (dma_cfg->source_data_size != dma_cfg->dest_data_size) {
+		return -EINVAL;
+	}
+
+	if (dma_cfg->dest_data_size != 1 && dma_cfg->dest_data_size != 2 &&
+	    dma_cfg->dest_data_size != 4) {
+		LOG_DBG("dest_data_size must be 1, 2, or 4 (%" PRIu32 ")", dma_cfg->dest_data_size);
+		return -EINVAL;
+	}
+
+	if (dma_cfg->channel_direction > PERIPHERAL_TO_MEMORY) {
+		LOG_DBG("channel_direction must be MEMORY_TO_MEMORY, "
+			"MEMORY_TO_PERIPHERAL or PERIPHERAL_TO_MEMORY (%" PRIu32 ")",
+			dma_cfg->channel_direction);
+		return -ENOTSUP;
+	}
+
+	data->channels[channel].config = dma_channel_get_default_config(channel);
+
+	data->channels[channel].source_address = (void *)dma_cfg->head_block->source_address;
+	data->channels[channel].dest_address = (void *)dma_cfg->head_block->dest_address;
+	data->channels[channel].block_size = dma_cfg->head_block->block_size;
+	channel_config_set_read_increment(&data->channels[channel].config,
+					  dma_cfg->head_block->source_addr_adj ==
+						  DMA_ADDR_ADJ_INCREMENT);
+	channel_config_set_write_increment(&data->channels[channel].config,
+					   dma_cfg->head_block->dest_addr_adj ==
+						   DMA_ADDR_ADJ_INCREMENT);
+	channel_config_set_transfer_data_size(
+		&data->channels[channel].config,
+		dma_rpi_pico_transfer_size(dma_cfg->source_data_size));
+	channel_config_set_dreq(&data->channels[channel].config,
+				RPI_PICO_DMA_SLOT_TO_DREQ(dma_cfg->dma_slot));
+
+	data->channels[channel].callback = dma_cfg->dma_callback;
+	data->channels[channel].user_data = dma_cfg->user_data;
+	data->channels[channel].direction = dma_cfg->channel_direction;
+
+	return 0;
+}
+
+static int dma_rpi_pico_reload(const struct device *dev, uint32_t ch, uint32_t src, uint32_t dst,
+			       size_t size)
+{
+	const struct dma_rpi_pico_config *cfg = dev->config;
+	struct dma_rpi_pico_data *data = dev->data;
+
+	if (ch >= cfg->channels) {
+		LOG_ERR("reload channel must be < %" PRIu32 " (%" PRIu32 ")", cfg->channels, ch);
+		return -EINVAL;
+	}
+
+	if (dma_channel_is_busy(ch)) {
+		return -EBUSY;
+	}
+
+	data->channels[ch].source_address = (void *)src;
+	data->channels[ch].dest_address = (void *)dst;
+	data->channels[ch].block_size = size;
+	dma_channel_configure(ch, &data->channels[ch].config, data->channels[ch].dest_address,
+			      data->channels[ch].source_address, data->channels[ch].block_size,
+			      true);
+
+	return 0;
+}
+
+static int dma_rpi_pico_start(const struct device *dev, uint32_t ch)
+{
+	const struct dma_rpi_pico_config *cfg = dev->config;
+	struct dma_rpi_pico_data *data = dev->data;
+
+	if (ch >= cfg->channels) {
+		LOG_DBG("start channel must be < %" PRIu32 " (%" PRIu32 ")", cfg->channels, ch);
+		return -EINVAL;
+	}
+
+	if (dma_channel_is_busy(ch)) {
+		LOG_DBG("channel %" PRIu32 " is busy", ch);
+		return -EBUSY;
```

Comment:
I removed this `busy` check.

As a result, in case of the dma_start call on transfer ongoing, set new values to the DMA register, but continue transferring.
It looks the same as ignoring.

The behavior does not violate the API description, so it is no problem. Could you please check at this point?

Reply:
I don’t have a pi pico to check but I believe your description suffices.